core/vm/evmc.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/evmc/bindings/go/evmc"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type EVMC struct {
+	instance *evmc.Instance
+	env      *EVM
+	readOnly bool // TODO: The readOnly flag should not be here.
+}
+
+var (
+	createMu     sync.Mutex
+	evmcConfig   string // The configuration the instance was created with.
+	evmcInstance *evmc.Instance
+)
```

Comment:
This global instance seems very dangerous long term (e.g. what happens if I want both execution and tracing with different options running concurrently? or block processing and mining concurrently? or concurrent transaction execution?).

How heavy is creating an EVMC instance? Do we have some numbers on it? Can't we create it on the fly when needed? Perhaps we could have a pool of them if they are expensive?

Reply:
Ideally if we could have some sort of a `map[config][]evmcInstance` (just an example for the idea). Then we could cover multiple code paths executing either the same or different EVMC instances.

Reply:
The simplest is possibly a `sync.Map` (thread safe) containing `sync.Pool` objects (thread safe + auto GC) containing `evmc.Instance` objects.

Reply:
I'm not a fan of global objects either, but it might be unavoidable in this case.

Let me start with some overview. First of all, the cost of creating an EVMC instance for interpreters is literally 0. They simply return their name and the ABI version. This is more expensive for EVMJIT where LLVM requires some global initialization to be performed. But anyway, the `evmcInstance` is something else than an instance of thje `Interpreter`. The `evmcInstance` does not have any execution context attached to it. The execution context of a VM is hidden and handled internally by the `Execute()` method which can be called concurrently. This also mean that e.g. you cannot inspect if an EVMC instance is in static mode currently.  It simply represents a handle of dynamically loaded EVMC module.

That's why I think this is unavoidable for DLLs (you don't want to call dlopen for every contract execution) especially if they are configured with CLI options. There will be some maintenance cost related to that.

Configuration options for EVMC make it even more complicated. However, I'm leaning toward a decision to allow passing EVMC options only when an EVMC instance is created. If you need different options you need to create another instance. Moreover, tracing will be probably handled differently (not by EVMC options).

In this PR, you can only create single EVM and single EWASM instances so we have 2 different instances at most. I can try to use `sync.Map` as a start (it should at least reduce the number of `panic()` occurrences). Let me know what you think about it.